### PR TITLE
Implement message splitting and merging

### PR DESCRIPTION
**Pull request by ColdenCullen**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: include/aws/mqtt/private/packets.h [hunk 1]
```diff
@@ -209,6 +209,9 @@ int aws_mqtt_packet_publish_init(
 AWS_MQTT_API
 int aws_mqtt_packet_publish_encode(struct aws_byte_buf *buf, const struct aws_mqtt_packet_publish *packet);
 
+AWS_MQTT_API
+int aws_mqtt_packet_publish_encode_no_payload(struct aws_byte_buf *buf, const struct aws_mqtt_packet_publish *packet);
```
Comment (justinboswell) [inline include/aws/mqtt/private/packets.h]: Couldn't this more clearly be named aws_mqtt_packet_publish_encode_headers()?

File: source/client_channel_handler.c [hunk 2]
```diff
@@ -250,31 +270,96 @@ static int s_process_read_message(
...
+        if (result) {
+            goto handle_error;
+        }
+
+        /* If the packet is still incomplete, don't do anything with the data. */
+        if (!packet_complete) {
+            return AWS_OP_SUCCESS;
+        }
+
+        /* Handle the pending packet */
```
Comment (justinboswell) [inline source/client_channel_handler.c]: Handle the completed pending packet

File: source/client_channel_handler.c [hunk 3]
```diff
@@ -250,31 +270,96 @@ static int s_process_read_message(
...
+        AWS_ZERO_STRUCT(packet_header);
+        int result = aws_mqtt_fixed_header_decode(&header_decode, &packet_header);
+
+        /* Calculate how much data was read. */
+        const size_t fixed_header_size = message_cursor.len - header_decode.len;
+
+        if (result) {
+            if (aws_last_error() == AWS_ERROR_SHORT_BUFFER) {
+                /* Message data too short, store data and come back later. */
+                connection->pending_packet.capacity = fixed_header_size + packet_header.remaining_length;
```
Comment (justinboswell) [inline source/client_channel_handler.c]: This is a bit odd, writing to capacity like this, why not just pass the computed size into aws_byte_buf_init() directly, or store it in a local first?

File: tests/aws_iot_client_test.c [hunk 4]
```diff
@@ -45,8 +46,10 @@ AWS_STATIC_STRING_FROM_LITERAL(s_client_id, "aws_iot_client_test");
 AWS_STATIC_STRING_FROM_LITERAL(s_subscribe_topic, "a/b");
 AWS_STATIC_STRING_FROM_LITERAL(s_hostname, "a1ba5f1mpna9k5-ats.iot.us-east-1.amazonaws.com");
 
-static uint8_t s_payload[] = "This s_payload contains data. It is some good ol' fashioned data.";
-enum { PAYLOAD_LEN = sizeof(s_payload) };
+enum { PUBLSIHES = 20 };
```
Comment (justinboswell) [inline tests/aws_iot_client_test.c]: PUBLISHES

File: tests/aws_iot_client_test.c [hunk 5]
```diff
@@ -219,23 +243,29 @@ int main(int argc, char **argv) {
...
             AWS_MQTT_QOS_AT_LEAST_ONCE,
             false,
             &payload_cur,
-            NULL, // &s_mqtt_publish_complete,
-            (void *)payload);
+            &s_on_puback,
+            &args);
 
-        sleep(1);
+        /* Don't throttle me daddy */
```
Comment (justinboswell) [inline tests/aws_iot_client_test.c]: /* prevent service endpoint message throttling */ :)

File: tests/aws_iot_client_test.c [hunk 6]
```diff
@@ -219,23 +243,29 @@ int main(int argc, char **argv) {
...
+        if (i % 100 == 0) {
+            sleep(1);
```
Comment (graebm) [inline tests/aws_iot_client_test.c]: avoid sleep on the 1st packet
`((i > 0) && (i % 100 == 0))`

File: source/client_channel_handler.c (hunk 3 above)
Comment (ColdenCullen) [reply source/client_channel_handler.c]: I think this is the by-product of a few refactorings. Will merge.

**Pull request merged**